REGRESSION (252580@main): [ macOS wk1 ] imported/w3c/web-platform-tests/cookies/value/value-ctl.html is a flaky failure

cdumez opened the pull request:
#### 6af2fb740dc255b0b19f46138ef0a545e7084619
<pre>
REGRESSION (252580@main): [ macOS wk1 ] imported/w3c/web-platform-tests/cookies/value/value-ctl.html is a flaky failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=245014">https://bugs.webkit.org/show_bug.cgi?id=245014</a>
rdar://99766248

Reviewed by Darin Adler.

Fix testRunner.removeAllCookies() so that:
1. It takes a completion handler which only gets called after the cookies have been removed,
   which usually happens asynchronously.
2. It also clear the cookie cache in the WebProcess

This addresses the flakiness and makes the test pass consistently on Cocoa ports.

* LayoutTests/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(window.test_driver_internal.delete_all_cookies):
* LayoutTests/platform/ios-15/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac-bigsur/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::deleteAllCookies):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::forEachProcessForSession):
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcessesForSession):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::deleteAllCookies):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
(+[WebPreferences _clearNetworkLoaderSession:]):
(+[WebPreferences _clearNetworkLoaderSession]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebPreferencesPrivate.h:
* Tools/DumpRenderTree/TestRunner.cpp:
(removeAllCookiesCallback):
* Tools/DumpRenderTree/TestRunner.h:
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(setWebPreferencesForTestOptions):
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
(TestRunner::removeAllCookies):
* Tools/DumpRenderTree/win/TestRunnerWin.cpp:
(TestRunner::removeAllCookies):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::removeAllCookies):
(WTR::TestRunner::callRemoveAllCookiesCallback):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveSynchronousMessageFromInjectedBundle):
(WTR::TestController::removeAllCookies):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didRemoveAllCookies):
* Tools/WebKitTestRunner/TestInvocation.h:

Canonical link: <a href="https://commits.webkit.org/256865@main">https://commits.webkit.org/256865@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4d1876be80458e00a4cfe1203e7002d0b0cc920

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30141 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106550 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166826 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101000 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6524 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35025 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89418 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103244 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102697 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83641 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31915 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86746 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/328 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/310 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21522 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4749 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5101 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1544 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40827 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
```diff
@@ -116,10 +118,27 @@ void HTTPCookieStore::deleteCookie(const WebCore::Cookie& cookie, CompletionHand
+                if (!process.canSendMessage())
+                    return;
+                auto pid = process.processIdentifier();
+                WTFLogAlways("CHRIS: sendWithAsyncReply WC BEGIN - PID: %d", pid);
```
You probably want to take these out; there are multiple in more than once source file.

cdumez replied:
Yes, sorry about this. I really miss webkit-patch's behavior that would show me a pretty-diff before uploading.

darinadler replied:
I find that I always have to open a terminal window next to the window where I am composing my change log and carefully read the diff

But of course, I am doing that to compose my per-function comments, which apparently are not something other WebKit contributors want to require! That process is where I notice so many mistakes I have made

darinadler commented on Tools/DumpRenderTree/TestRunner.cpp:
```diff
@@ -268,8 +268,11 @@ static JSValueRef removeAllVisitedLinksCallback(JSContextRef context, JSObjectRe
 
 static JSValueRef removeAllCookiesCallback(JSContextRef context, JSObjectRef function, JSObjectRef thisObject, size_t argumentCount, const JSValueRef arguments[], JSValueRef* exception)
 {
+    if (argumentCount < 1)
+        return JSValueMakeUndefined(context);
```
I think we’d want to throw an exception in this case rather than silently doing nothing, but I understand this is just test machinery so maybe this is OK.

cdumez replied:
I agree this isn't great. However, this is the pattern in this file. Every function just returns early with undefined if they don't get enough parameters. I couldn't even find one that throws an exception.

cdumez committed:
REGRESSION (252580@main): [ macOS wk1 ] imported/w3c/web-platform-tests/cookies/value/value-ctl.html is a flaky failure
https://bugs.webkit.org/show_bug.cgi?id=245014
rdar://99766248

Reviewed by Darin Adler.

Fix testRunner.removeAllCookies() so that:
1. It takes a completion handler which only gets called after the cookies have been removed,
   which usually happens asynchronously.
2. It also clear the cookie cache in the WebProcess

This addresses the flakiness and makes the test pass consistently on Cocoa ports.

* LayoutTests/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/resources/testdriver-vendor.js:
(window.test_driver_internal.delete_all_cookies):
* LayoutTests/platform/ios-15/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac-bigsur/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac-monterey/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt: Removed.
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::deleteAllCookies):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::forEachProcessForSession):
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcessesForSession):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::deleteAllCookies):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
(+[WebPreferences _clearNetworkLoaderSession:]):
(+[WebPreferences _clearNetworkLoaderSession]): Deleted.
* Source/WebKitLegacy/mac/WebView/WebPreferencesPrivate.h:
* Tools/DumpRenderTree/TestRunner.cpp:
(removeAllCookiesCallback):
* Tools/DumpRenderTree/TestRunner.h:
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(setWebPreferencesForTestOptions):
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
(TestRunner::removeAllCookies):
* Tools/DumpRenderTree/win/TestRunnerWin.cpp:
(TestRunner::removeAllCookies):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::removeAllCookies):
(WTR::TestRunner::callRemoveAllCookiesCallback):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveSynchronousMessageFromInjectedBundle):
(WTR::TestController::removeAllCookies):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
(WTR::TestInvocation::didRemoveAllCookies):
* Tools/WebKitTestRunner/TestInvocation.h:

Canonical link: https://commits.webkit.org/256865@main
